src/Common/Utilities/OAuthHelper.cs

```diff
@@ -0,0 +1,389 @@
+
+#endregion
+
+using JetBrains.Annotations;
+
+namespace Zeiss.IMT.PiWeb.Api.Common.Utilities
+{
+	#region usings
+
+	using System;
+	using System.Collections.Generic;
+	using System.IdentityModel.Tokens.Jwt;
+	using System.IO;
+	using System.Linq;
+	using System.Security.Claims;
+	using System.Security.Cryptography;
+	using System.Threading.Tasks;
+	using IdentityModel;
+	using IdentityModel.Client;
+	using Zeiss.IMT.PiWeb.Api.OAuthService;
+
+	#endregion
+
+	public static class OAuthHelper
+	{
+		#region constants
+
+		private const string ClientId = "f1ddf74a-7ed1-4963-ab60-a1138a089791";
+		private const string ClientSecret = "d2940022-7469-4790-9498-776e3adac79f";
+		private const string RedirectUri = "urn:ietf:wg:oauth:2.0:oob";
+
+		private static readonly string _CacheFilePath = Path.Combine( Environment.GetFolderPath( Environment.SpecialFolder.ApplicationData ), @"Zeiss\PiWeb\OpenIdTokens.dat" );
+
+		#endregion
+
+		#region members
+
+		// FUTURE: replace by a real cache with cleanup, this is a memory leak for long running processes
+		private static readonly CredentialRepository _AccessTokenCache = new CredentialRepository(_CacheFilePath);
+
+		#endregion
+
+		#region methods
+
+		public static JwtSecurityToken DecodeSecurityToken( string jwtEncodedString )
+		{
+			return new JwtSecurityToken( jwtEncodedString );
+		}
+
+		public static string IdentityClaimsToFriendlyText( IList<Claim> claims )
+		{
+			var name = claims.SingleOrDefault( claim => claim.Type == "name" )?.Value;
+			var email = claims.SingleOrDefault( claim => claim.Type == "email" )?.Value;
+
+			return $"{name} ({email})";
+		}
+
+		public static string TokenToFriendlyText( string jwtEncodedString )
+		{
+			if ( string.IsNullOrEmpty( jwtEncodedString ) )
+				return "";
+
+			try
+			{
+				var decodedToken = DecodeSecurityToken( jwtEncodedString );
+				if ( decodedToken != null )
+				{
+					return IdentityClaimsToFriendlyText( decodedToken.Claims.ToList() );
+				}
+			}
+			catch
+			{
+				// ignored
+			}
+			return "";
+		}
+
+		public static string TokenToMailAddress( string jwtEncodedString )
+		{
+			if( string.IsNullOrEmpty( jwtEncodedString ) )
+				return "";
+
+			try
+			{
+				var decodedToken = DecodeSecurityToken( jwtEncodedString );
+				if( decodedToken != null )
+				{
+					return decodedToken.Claims.SingleOrDefault( c => string.Equals( c.Type, "email" ) )?.Value;
+				}
+			}
+			catch
+			{
+				// ignored
+			}
+			return "";
+		}
+
+		private static string GetInstanceUrl( [NotNull] string databaseUrl )
+		{
+			if( databaseUrl == null )
+				throw new ArgumentNullException( nameof( databaseUrl ) );
+
+            databaseUrl = databaseUrl.Replace( "/DataServiceSoap", string.Empty );
+            databaseUrl = databaseUrl.Trim('/');
+            return databaseUrl;
+        }
+
+		private static OAuthTokenCredential TryGetCurrentOAuthToken( string instanceUrl, ref string refreshToken )
+        {

```

Comment:
Die Mischung von Tabs und Spaces ist hier auch schlecht zu lesen. Das sollte mit einmal Code-Formatierung anwenden auch geloest sein.

Auch hier sollte in Zukunft die .editorconfig helfen.